### DISCUSSION TITLE
Usage of Foreign Key Constraints

### DISCUSSION THREAD

**Discussion by ghislaineguerin** [mathesar-foundation/mathesar#744] [Maintainer discussions]
The design for issue #243  is ready for review. Please follow the [Review Guidelines](https://wiki.mathesar.org/en/design/process/review-guidelines) and make sure to review before the set deadline expires.

### Document
https://wiki.mathesar.org/en/design/specs/foreign-key-constraints

**The document includes a walkthrough video**

### Reviewers
@kgodey @pavish @seancolsen @mathemancer  Your review is required.
Feedback from the community is also welcome. Contributors can find more context about Mathesar's design strategy [here](https://wiki.mathesar.org/design).

### Deadline
The deadline for submitting feedback is set 2 days from now, on Tuesday, Oct 26.

**Comment by seancolsen** [comment]
## I like

- The styling of the values in the FK column, setting them apart from other columns.
- The functionality of the "Lookup records" editor, generally.
- The idea of an "Add Record" button if no records are found -- though I do wonder what happens after I click the button. We'll need to think that through in more detail, though probably not right away.

## Suggested changes

### Use a modal for adding/editing columns

The workflow of adding a new column with a generated named an an assumed type feels a bit awkward to me. I'd rather use a modal for the purpose of creating and editing a column. Here's a [prototype](https://svelte.dev/repl/22d489f8479547b2adc8ee8edf0fe9b9?version=3.44.0) which demonstrates what I have in-mind for this "Add/Edit Column" modal;

![Peek 2021-10-22 15-45](https://user-images.githubusercontent.com/42411/138514195-72efd8b2-11e9-429e-932d-42c7e6207bbd.gif)

- That demo is for _adding_ a column, but it would be very similar when editing a column and would allow the user to add an FK to an existing column or to remove an FK constraint while keeping the column intact.
- Benefits of this approach:
    - The type of a new FK column can be set automatically from the target column.
    - The name of a new FK column can be pre-filled for the user based on the target table and column.
    - It safeguards against the case where a user forgets to customize the column name.
    - It helps steer the user towards thinking more intentionally about the types of their columns.
    - Adding the modal would allow us to simplify the column header dropdown menu (likely to be a very commonly-used menu), by removing the (likely infrequently used) options for "Link to another table" and "Remove link".

### Use a link icon instead of a key icon

In the column header for a FK column, I like the addition of an icon, but I think the key will be misleading to non-technical users. Something like [faLink](https://fontawesome.com/v5.15/icons/link?style=solid) better illustrate the concept.

## Ancillary thoughts

- There's one aspect of this FK workflow that I think really needs to eventually be improved (but probably not at this phase of the development process) -- when the user is viewing or editing the value within a FK column, I think it's important that we display other fields beyond just the FK target column. If I've set `track.artist_id` to reference `artist.id`, I'd like to search by `artist.name` and see `artist.name` most of the time. That additional requirement presents a very complex problem UX though, an not one I think we should rush into solving. Once we're further along, I think we might have a great opportunity to leverage our "views" feature in building the FK selector and display. For example, I could designate a view to be used when selecting an artist and another view to be used when displaying a linked artist.

- Here are some aspects of the prototype that I found confusing but that I'm overlooking because I'm guessing they're artifacts of the rapid prototyping process and irrelevant to this FK workflow discussion. (Just mentioning here in case I'm missing something.)
    - I would not expect the `artists` table to contain a column titled `releaseId` because that implies the an artist can have no more than one release. Same for `artists.trackId`.
    - "Link to multiple" button -- I'm not sure what's supposed to happen after I click on a table within the modal. I have some more thoughts about this topic, but it seems like the "Link to Multiple" scenario is beyond the scope of this spec, so I'll hold off on elaborating for now.

- As a sort of meta-comment, I found the format of the [specs](https://wiki.mathesar.org/en/design/specs/foreign-key-constraints) confusing. After reading "Scenarios", I assumed that each heading underneath would be one distinct scenario that could occur in isolation. Now I realize that the whole page is one scenario, with many different steps that happen in sequence. For future specs I'd suggest changing "Scenarios" to "Steps" and numbering each of the subheading sequentially. This is a minor point, but I mention it for the purpose of honing our process, especially since it seems like we're soon to welcome two more devs to the team who will also need to get acquainted with these processes.

**Reply by ghislaineguerin** [reply]
Thanks for the feedback @seancolsen. I have made some updates based on your input. Here are some thoughts:

### Using a modal for new column creation 
I think you make some excellent points for combining column creation with table linking. I remember this was considered when we worked on the column options and ended up separating both. I believe it was because the foreign key wasn't necessarily a data type. We wanted to clarify that we'd still display the linked field data type and prevent users from changing it, thus blocking them from seeing this menu or seeing it as read-only.
Separating table relationships in a different menu could potentially be helpful in the future when we want to provide the user with suggestions on actions they could perform based on data contained in a column. We might detect, for example, that a column contains unique values that match those of another column and suggest they link them. If those concepts are separated, it will be easier to suggest specific actions, such as multiple record linking, which might need a separate flow.

### Using a link icon
We also considered this, but the URL data type already uses a link. I will give it some thought and consider other options. Indeed, us calling this a 'link' while showing a 'key' is confusing.

### Displaying other fields beyond the FK target column
You are right about the need to search records by other fields beyond the linked one. We have the concept of a lookup column. I have updated the document to explain this. By default, we assign the first column after 'id' as a lookup field.

### Link to multiple
Sorry, that was a leftover from another exploration. I have updated the prototype to remove it. Same for the releaseId column.

### Regarding specs
Agree. I described the scenario as the user wanting to link two tables but didn't specify the steps. Thanks for the suggestion.

**Reply by seancolsen** [reply]
> the URL data type already uses a link

Ah! I see.

- Maybe [external-link_alt](https://fontawesome.com/v5.15/icons/external-link-alt?style=solid) for an FK column?
- Or maybe [external-link_alt](https://fontawesome.com/v5.15/icons/external-link-alt?style=solid) or [code](https://fontawesome.com/v5.15/icons/code?style=solid) or [globe](https://fontawesome.com/v5.15/icons/globe?style=solid) for the URL data type?

**Comment by mathemancer** [comment]
### Setting up a relationship

You're switching back and forth between 'link' and 'relationship' both in spoken word, and in the iconography and words used in the UI.  To add to the confusion, "Relation" in "Relational Database Management System" refers to a _table_ (or view, or sequence; basically anything with columns).  A relationship between tables (as you're using it) is just one special case of a relation, represented by a table.  This is the source of 50 years confusion (basically, 'relation' is a fussy mathematical term that made perfect sense to the mathematician who invented the concept in the 1970s).  The common use of the term "relationship" comes from the informal concept of entity/relationship modeling, and high-quality resources dealing with RDBMSs will not use that term for describing links between tables (since relationships between relations is kind of nightmare terminology).  

I suggest sticking very fastidiously with 'link' to describe a link between two tables (or choosing some other term besides 'relationship').  I recognize this is less common terminology, but "relationship" is fraught with confusion in this space, and I'd really like to avoid having our product add to rather than subtract from that confusion.

### Lookup fields

Can we expand this UI for multi-column foreign keys?  I realize that's not the current design goal, but the design should be able to handle that eventually.  Also, we need to either handle the case where the user wants a non-distinct column for a lookup column, or guide them towards using only distinct columns for that purpose (I prefer the latter).  The reason is that it'll probably be impossible for them to figure out from the current UI which row they're choosing when looking at a list of IDs and non-distinct row entries (suppose it's track_length in the example from the video).

### Foreign Keys

I'd rather describe this as a constraint (and think of it that way) whereby the constrained table column(s) can hold only values which appear in the referenced column(s).  The reference is _from_ the table with the foreign key _to_ the other table.

### Scenarios

I think these look great overall.  Some notes:
- see above about relationship vs. link verbiage
- we should think about how to deal with non-distinct choices for lookup fields
- We should try to figure out how to portray the fact that the lookup field is just a stand in for the ID of a row; this is crucial and not that clear (to me).  I can envision a scenario where a user decides to use some oddball field (e.g., track length in the example), and ends up with the idea that they're connecting track lengths to artists or albums, not connecting tracks to artists or albums.  There's some verbiage in the write-up about this, but I don't see how it's portrayed in the videos (maybe I missed it).

**Comment by kgodey** [comment]
@ghislaineguerin I've already given you feedback on this while you were designing it, I don't have anything to add to the feedback that others have already given you. It looks good!

I did want to summarize our DM conversation on Matrix today in case anyone else had thoughts. The idea of lookup fields might be confusing when combined with being able to set the column to which you are creating a foreign key link. e.g. the scenario that @mathemancer wrote out. I think we might want to remove the concept of lookup fields for the first version of the product and instead just display a non-configurable preview of the other table, regardless of which column is selected as the foreign key.

**Comment by pavish** [comment]
## Adding the FK link:

### To an existing column
The current flow to configure existing columns as foreign keys look good to me. However we need loading and failure states to be defined on the UX spec.

### Adding a new column and then configuring it as a foreign key.

This is tricky. When linking a column to another column from a different table, the db types have to match.

* Consider the columns belong to the number type (which is a combination of different internal db types, that the user can only set going to advanced section).
* The referenced table has it as a float db type.
* The user adds a new column which creates a text column directly, and the user switches it to number (assume the default is integer).
* The user tries to link the new column to the other table.
* The user will get an error saying that they cannot link columns because types mismatch.
* They will also not be able to easily identify the cause in a single glance at the table.

We will need to consider how to make this easier for the user.

## Lookup fields:

We would need the UX to be updated to allow more than one lookup field. It might be better if we can choose the lookup field from the new fk column, rather than configuring it on the referenced table. 

For eg., in the dropdown to choose a value, it would be better if we provided an option to view and search by multiple columns. We woudn't have to explicitly go to the other table and configure the lookup columns.

This would satisfy @mathemancer's concerns.

## About adding new column (general)
About @seancolsen 's comments on changing the UX for adding column, I do think we need to consider having explicit actions to getting column info (especially the type and name) to be defined before the column is created. This is not specific to this design issue, however this was one of the items planned to have been discussed in the UI/UX sync. I'll mention it here since we are building on top of existing spec, and delaying the UI/UX sync to later milestones.

Currently as per the spec, column is supposed to be created with a default name and type, and we edit the column header direclty to change it. There is also no explicit save button. This introduces certain complexities during development.
* The backend does not have a static id for a column (right now, we have index, we discussed to make it index + name).
* Editing the column in-place without an explicit save button would require debouncing requests and sending save requests as the user types.
  * This may lead to multiple requests to the server, while the user is still typing and there may be delays in the server processing some of these requests. 
  * It could lead to the column being renamed on the backend, but the user is still typing. The client does not receive the updated name.
  * Since the requests to the column endpoint itself might depend on index + name, all further requests may fail.
  * This may seem like an edge case, but I usually run with network throttle enabled while developing and these kind of issues happen very often. (@seancolsen we have the CancellablePromise to prevent such issues on the client, however we can only abort the request on the client. The server would process the debounced requests as usual.)
  * These errors can be mitigated by using an explicit save button.
* The previously mentioned point on FK linking arises because we provide a default value.

All these would be resolved with an explicit save button.

Our current UI for adding columns shows a small dropdown which gets the name from the user. I propose using a dropdown to get the values while creating columns instead of the modal, to make the UI less intrusive.

**Comment by ghislaineguerin** [comment]
@kgodey @mathemancer @pavish @seancolsen Thanks for the feedback. I believe, based on the number of potential failure/confusion scenarios, that the best approach is the following:

## Restricting table-level functionality
In Mathesar, users will have tables that they want to link to display columns from both, create aggregations, and more. For this, we suggest that they use foreign keys to set those relationships. However, this concept is **too detached from the actual output** they have in mind, leading to confusion.

**Scenarios that users could face when dealing with foreign key columns:**

- The user wanted to add many fields from table B to table A, so instead, they added multiple foreign keys or references to each of those columns.
- The user picks a non-unique value column from table B to use as a foreign key in table A
- The user wanted to link multiple records from B to A and has set the relationship in the wrong table.

Consider the additional problems that could arise when users want to summarize records from table B in a column added to table A.

**There is no "simple" design solution for those edge cases unless the system takes care of foreign key management based on the actual user's needs rather than their initial and potentially unreliable intentions. The other option is to educate users on foreign key usage, which adds friction between the product and their immediate goals.**

For this reason, and after exploring most options I could think of, I believe restricting table functionality is the best one. By restricting, we keep the foreign key functionality accessible (existing databases might have them) but only from the 'table constraints' menu. So that advanced users can add or delete them at will. However, when approaching the needs of non-advanced users, we will set these based on outcomes they want to obtain in the context of editing or setting up a new view.

- From tables, users will add columns, populate data, set data types.
- From views, users will add columns from other tables directly from the 'add column' menu. They can also do this when creating a view from scratch, where the option to select more than one table will be available.
- The system can then 'auto-detect potential relationships based on shared values across tables. If none is found, the system can automatically create a foreign key field in one of the tables.

This approach would help hide the complexity and reduce the potential failure scenarios while still providing advanced users the ability to modify table constraints.

#### For example, in the absence of shared values across tables:

- A user adds a new view, and it's prompted by the system to select the tables that will be part of this view. - The user first selects the 'artist' table and a column or multiple from that table. Once they do this, a preview of the view's first three rows of the new view is displayed.
- The user then selects a second table, this time, they pick 'track' and choose a column to display.
- The system could not detect shared values between the tables. The preview shows a table with 'artists' listed while the 'track' column is empty.
- The system notifies the user about the lack of a shared values column and offers the option to "Create a reference column in 'track'" The user understands that the system will update the view dynamically once they start populating that column in the 'track' table.
- The user accepts the suggestion, creating a new view with both 'artist' and 'track' columns. The 'track' table now contains a new foreign key field that links to 'artist.'
- The user can edit that new field from the view or open the table. They can switch between table and view, so the impact of their actions is clear.

#### And in the case where there are shared values:
If there are shared values in selected tables, then it becomes much more manageable. The system will also create a foreign key constraint in the table to prevent problems with data type mismatch. For example, if 'track' has a string type 'ISRC' (International Standard Recording Code) that matches values in an 'ISRC' field from 'artist' set to 'numerical.' It could still create a 'foreign key' column and populate the Id values from the artist table.
When the system creates foreign key relationships, it will always default to the other table's primary key as the constrained values.

**Reply by kgodey** [reply]
@ghislaineguerin If I can sum up the idea, it seems like for the first version, we will:
- only be using the constraints view to add/edit foreign keys from Tables
- only link tables together by adding columns to Views and setting up relationships as needed

I like this idea, it simplifies things and it makes it clear to users that they should mainly be working with Views. 

However, we'll still need to figure out how to display foreign key columns in tables (regardless of how they are added).
- Will we still have lookup fields, or will we display a preview of the other table?
- How will we display multi-columns foreign keys?
- How will the user edit data in foreign key columns within tables?

**Reply by seancolsen** [reply]
If I'm interpreting correctly, it means that the only way I can set up an FK column is to create a view. That seems like a cumbersome workflow. For example, let's pretend I start with an empty database. I create an `artist` table with `id` and `name`. I add one artist. Now I want to add one album for that artist. Before I add the album I need an `album` table. To create the table in a way the allows me to _select_ an artist (instead of manually entering the artist id), I need an `artist_id` FK column. To add the column, I need a view. To add the view I need the table already. Dang! Am I misinterpreting?

**Reply by seancolsen** [reply]
There are so many difficult UX questions with this FK stuff that I think our discussion and decision-making process would benefit from narrowing the scope of the spec a bit. I'd be inclined to focus only on the following topics in this order:

1. Editing schema (i.e. adding/editing/deleting columns/constraints)
1. Viewing data (i.e. what do we display in the cell of an FK column -- and how does the user configure it)
1. Editing data (i.e. what's the UX for changing values within an FK column)

I recognize that these topics affect each other somewhat so it's not entirely possible to consider them in isolation. But I think we might come to decisions more easily if we can divide the problem up a bit.

**Reply by mathemancer** [reply]
To address the failure scenarios above, it occurs to me that the central confusion there is about what a foreign key is.  It is _not_ a display of a column from a different table in your table.  As @ghislaineguerin points out, the correct tool for viewing columns from multiple tables together is a view, not foreign key columns.  The value of the foreign key columns is that they let you define such a view quickly and easily by associating rows of two tables with each other in a well-defined way.

What a foreign key is is a link between two tables, each ideally representing some entity that says "artists have tracks" (and vice-versa), "albums have tracks" (and vice-versa), "artists have albums" (and vice-versa), etc.  The link is between tables, not columns.  The reason it's often portrayed as a constraint on a column is that this is the simplest way to make a robust link between two tables:  You choose one value from the primary key column of another table, and add that value to your current table.  This defines a function that lets you map from the row on your current table into a row on the other table, thereby linking those tables.

With that in mind, I suggest letting users choose to link a _table_ (with a primary key) to their current table, and set up a foreign key column for that purpose.  We should automatically set the foreign key column to hold primary key values from the other table, and make it obvious that when choosing a "value" for that column, you're choosing a _row_ from the other table to associate with the _row_ in the current table.

This would guide users towards better data modeling, and simplify the flow.  There wouldn't be questions about "which column to choose", "distinct columns", etc.  If they want something more complex, they can use the more advanced interface.

I do like the option of automatically creating a foreign key map between tables containing columns the user wants to have in the same view, but I don't think that should be the primary way to do it.  The reason is that they may _not_ want to view those columns together, but they may still want to be able to attach artists to tracks.  To me, it's not the desire to see those columns next to each other that should drive the foreign key mapping, but rather the desire to link those entities together logically.  (I.e., the "artist" and "track" entities).

**Reply by mathemancer** [reply]
W.r.t. @seancolsen's numbered list:  I completely agree that splitting into those three tasks is a good idea:
1.  I think the flow should essentially be:  "user wants to link to another table", "user chooses another table".  We should create the appropriate foreign key mapping in the background.  It might make sense to offer a choice of one-one, many-one, many-many, but that could be advanced.
2. This is the hardest part.  TBH, I think the view from the table should just show the ID at this point.  If they want to view other columns from the linked table, they can easily create a view using their fancy new foreign key mapping.
3. I think the editing (i.e., actually creating the row-by-row links) is easiest in a view.  This is related to the flow @ghislaineguerin described in the parent post.  I.e., they go to the view, and see the columns of their table, along with blank columns named for the other table.  They start filling in those columns, with each value restricted to the appropriate choices.  Once they have enough values filled for a definition of which row they wanted in the other table, the rest automatically populates.  (Under the hood, of course, we've simply filled in the foreign key column with the appropriate primary key value.)  We could also do some sort of drop-down in the column, but I think that will be quite messy UI-wise.

**Reply by ghislaineguerin** [reply]
@mathemancer I agree that the table should only show ID, other solutions open up possibilities for confusion. 
I also agree with @seancolsen process proposal.

**Reply by seancolsen** [reply]
@mathemancer wrote:

> I suggest letting users choose to link a table (with a primary key) to their current table, and set up a foreign key column for that purpose.

Very much agree.

That's exactly the thoughts I had when [proposing](https://github.com/centerofci/mathesar/discussions/744#discussioncomment-1523164) the modal [Add/Edit Column](https://svelte.dev/repl/22d489f8479547b2adc8ee8edf0fe9b9?version=3.44.0) interface. Within that interface, we could even decide to _force_ the user to use the primary key. It would certainly be an easier place to start.

**Reply by ghislaineguerin** [reply]
@seancolsen I get your point with the add/edit column, but the user will still end up with a column containing ID numbers and not the result they were expecting. The primary key for tables created within Mathesar will be set by default to the ID column. We still need, from this intermediate step, direct users to view so that they can display other columns from the linked table.

**Reply by seancolsen** [reply]
@ghislaineguerin I'm confused by:

> but the user will still end up with a column containing ID numbers and not the result they were expecting

Don't we _want_ to store the ID numbers? I think the question of what we _display within the table_ is a separate topic from the question of how the user sets up the column.

**Reply by ghislaineguerin** [reply]
@seancolsen if the user wanted a table that shows 'track name' and 'artist name,' this is an intermediate step, where they have track name and id of the artist, but not artist name. They still need to create a view to achieving their desired outcome. What you are proposing still reproduces the functionality of adding FK constraint to a column which we could do from the table constraints modal.
The idea is to not handle the foreign key functionality in the context of adding a column, as it is misleading for users.

**Reply by kgodey** [reply]
> If I'm interpreting correctly, it means that the only way I can set up an FK column is to create a view.

I don't think this part of @seancolsen's message was addressed. To clarify, that's not the _only_ way to set up an FK column, you can also use the table constraints modal.

**Reply by mathemancer** [reply]
> That's exactly the thoughts I had when [proposing](https://github.com/centerofci/mathesar/discussions/744#discussioncomment-1523164) the modal [Add/Edit Column](https://svelte.dev/repl/22d489f8479547b2adc8ee8edf0fe9b9?version=3.44.0) interface. Within that interface, we could even decide to _force_ the user to use the primary key. It would certainly be an easier place to start.

My point is that if the user is thinking in terms of columns when linking entities, we've already lost.  It would be preferable for them to be thinking "link artists to their tracks" (on the entity level) rather than "create a foreign key column in the tracks table pointing to artists.id" .  The specific columns used to create the mapping between the entities should be irrelevant, as long as it defines a proper function from the set of tracks to the set of artists.

**Reply by mathemancer** [reply]
One caveat:  The direction of the link is defined by which table gets the foreign key (I.e., a different mapping direction is implied if the foreign key is in the artists table rather than the tracks table).  This is one issue I have with the view setup for creating these (or at least a wrinkle to be smoothed) is to portray this asymmetry to the user.  Suppose the user creates a view that puts columns from `artists` next to columns from `tracks`.  How do we choose which table gets a foreign key, and how do we portray that choice and its consequences to the user?

This is one reason I still think this linking would be most easily done outside of the view, in the context of linking a table to another table.  As I think about it, asking the user which way they want to link (maybe with some decision tree help) may be crucial:
- "Can an album have more than one track?" (yes)
- "Can a track have more than one album?" (no)
-> set up a foreign key album_id in the tracks table

- "Can an album have more than one artist?" (yes)
- "Can an artist have more than one album?" (yes)
-> set up a many-to-many mapping with a mapping table

- "Can an artist have more than one profile pic URL?" (no; just for example)
- "Can a profile pic URL have more than one artist?" (no; just for example)
-> set up one-to-one (which gets the foreign key is irrelevant.  Could choose arbitrarily or ask user)

**Reply by seancolsen** [reply]
@kgodey wrote:

> you can also use the table constraints modal.

Thanks. That helps to address my concern.

**Reply by pavish** [reply]
> My point is that if the user is thinking in terms of columns when linking entities, we've already lost.  It would be preferable for them to be thinking "link artists to their tracks" (on the entity level) 

I really like this statement from @mathemancer. I think this makes a lot more sense and simplifies a bunch of issues around how we plan to link tables.

**Reply by seancolsen** [reply]
I know we've moved on from this design conversation now, but I've been marinating on this topic a bit more today and wanted to share some additional thoughts.

First off, I'll say that I support moving forward with the latest [specs in #243 (comment)](https://github.com/centerofci/mathesar/issues/243#issuecomment-953166891), and I don't want to propose a different approach.

But this statement from @mathemancer caught my eye too and I want to explore it more deeply.

> if the user is thinking in terms of columns when linking entities, we've already lost

I disagree. I _do_ think in terms ot columns when linking entities. And I've been assuming we'd also like to steer our users towards thinking of entities as linked by columns. Maybe there's something I'm missing about our product vision/goals or databases in general. If so, I'd like to understand so that I can get on the same page.

Here's an example of a schema which demonstrates why my mental model is uses columns to link entities instead of doing the linking "at the entity level".

```yaml
- address
  - id
  - street
  - city
  - ... 

- shipment
  - id
  - origin (FK to address.id)
  - destination (FK to address.id)
  - ...
```

In this case, thinking _"link addresses to their shipments"_ is not sufficient.

To be clear though, I'm just "thinking in terms of columns" on the "shipment" side of the link. Whereas on the "address" side of the link, I agree that it's great to "think in terms of entities", which means always setting up the FK to point at a primary key.

All this said, we can definitely continue to tweak the UX once the features are built. And I don't want to get too bogged down in debate. But if there's a fundamental concept that I'm missing, let me know.

**Reply by mathemancer** [reply]
While there's certainly room for reasonable minds to disagree, I would see this as "shipments link addresses to other addresses", i.e., the shipments table is an annotated mapping table for a many-to-many map.  I agree that this is a bit contrived, but I think the simplification generally leads to better modeling.  I suppose from your point of view, the "shipments" class has an attribute that's an "address" class, i.e., shipments are composed of addresses (and other stuff).  

Also, I'm _not_ of the opinion that every single data modeling problem can be reduced to mappings between table-sized entities.  I do think, however, that the exercise of really trying to accomplish that generally leads to better data models.

EDIT:  now I'm distracted trying to decide if I can think of a good example where the model is good, and both entities are composed of the other. (I.e., where column-thinking works in both directions)

**Reply by pavish** [reply]
> My point is that if the user is thinking in terms of columns when linking entities, we've already lost

I think this suits us better because it reflects how the database works internally. FK constraints are properties of the table and not of the column specifically. Structuring the UX with a similar principle would make it a lot easier, especially when we would have to deal with multi-column foreign keys in the future.

Having said all that, I think the questions apply only for someone who's not strong technically. We would need an advanced section in the future where the user can explicitly specify foreign keys and links in terms of columns, provided that it's within the table context and not within the column dropdown. 

Another way to think at this would be:
_Single column one to one/many FK_: Link a column to a record/records in a different table
_Multi column one to one/many FK_: Link a set of columns to a single record/records in a different table
_Many-to-many mapping_: Link a record/records to a record/records in a different table, which requires a mapping table

It's somewhat fuzzy where the context lies generally speaking. But on the UX standpoint, I think it simplifies a bunch of stuff if it does not lie in the column configuration, but rather on the table context.

**Reply by kgodey** [reply]
> I disagree. I do think in terms ot columns when linking entities. And I've been assuming we'd also like to steer our users towards thinking of entities as linked by columns.

I think that in most data modeling problems, the user is trying to link a column on one table to the entirety of another table i.e. you're not trying to link a column on the `shipment` table to a specific column on the `address` table; you're linking it to an entire `address` entity.  I think you could even think about `address` as being the data type of the `origin` column. Maybe we'd want to convey that concept to the user eventually.

However, the problem with starting from a column while creating links in the UX is that the correct set up of the link might not involve creating a column at all. For example, if a non-technical user is browsing the `address` table and wants to mark an address as the origin of a shipment, they might try to create a `shipments` column and go from there, which wouldn't work since an address can be associated with many shipments and the link should properly be created from the other side.

**Reply by seancolsen** [reply]
@kgodey said:

> they might try to create a `shipments` column and go from there

Really good point. Thanks for pointing that out. Very compelling example.

**Comment by ghislaineguerin** [comment]
@seancolsen @pavish @mathemancer @kgodey 
Thanks, everyone, for your input. @kgodey and I talked and agreed to address the foreign key issue according to the increments listed here: https://github.com/centerofci/mathesar/issues/243#issuecomment-953166891

**Reply by kgodey** [reply]
Since this discussion is getting long, we'll open a new one once the spec has been updated.
